processor/src_hydration_stage.go

```diff
@@ -0,0 +1,192 @@
+	archivalJobs                  []*jobsdb.JobT
+	eventSchemaJobsBySourceId     map[SourceIDT][]*jobsdb.JobT
+	connectionDetailsMap          map[string]*reportingtypes.ConnectionDetails
+	statusDetailsMap              map[string]map[string]*reportingtypes.StatusDetail
+	enricherStatusDetailsMap      map[string]map[string]*reportingtypes.StatusDetail
+	botManagementStatusDetailsMap map[string]map[string]*reportingtypes.StatusDetail
+	eventBlockingStatusDetailsMap map[string]map[string]*reportingtypes.StatusDetail
+	destFilterStatusDetailMap     map[string]map[string]*reportingtypes.StatusDetail
+	reportMetrics                 []*reportingtypes.PUReportedMetric
+	totalEvents                   int
+	groupedEventsBySourceId       map[SourceIDT][]types.TransformerEvent
+	eventsByMessageID             map[string]types.SingularEventWithReceivedAt
+	jobIDToSpecificDestMapOnly    map[int64]string
+	statusList                    []*jobsdb.JobStatusT
+	jobList                       []*jobsdb.JobT
+	sourceDupStats                map[dupStatKey]int
+	dedupKeys                     map[string]struct{}
+}
+
+func (proc *Handle) srcHydrationStage(partition string, message *srcHydrationMessage) (*preTransformationMessage, error) {
+	spanTags := stats.Tags{"partition": partition}
+	ctx, processJobsSpan := proc.tracer.Trace(message.subJobs.ctx, "srcHydrationStage", tracing.WithTraceTags(spanTags))
+	defer processJobsSpan.End()
+
+	if proc.limiter.srcHydration != nil {
+		defer proc.limiter.srcHydration.BeginWithPriority(partition, proc.getLimiterPriority(partition))()
+		defer proc.stats.statSrcHydrationStageCount(partition).Count(
+			lo.Sum(lo.MapToSlice(message.groupedEventsBySourceId, func(key SourceIDT, jobs []types.TransformerEvent) int {
+				return len(jobs)
+			})))
+	}
+
+	// Process sources in parallel using errgroup
+	g, ctx := errgroup.WithContext(ctx)
+
+	// Mutex to protect shared maps
+	var sharedMapsMutex sync.Mutex
+
+	for sourceId, jobs := range message.groupedEventsBySourceId {
+		g.Go(func() error {
+			source, err := proc.getSourceBySourceID(string(sourceId))
+			if err != nil {
+				return err
+			}
+			if !source.IsSourceHydrationSupported() {
+				return nil
+			}
+
+			hydratedJobs, err := proc.hydrate(ctx, source, jobs)
+			if err != nil {
+				proc.logger.Errorn("failed to hydrate source events, skipping jobs",
+					obskit.SourceID(string(sourceId)),
+					obskit.Error(err),
+				)
+			}
+
+			// Update shared maps with mutex protection
+			sharedMapsMutex.Lock()
+			defer sharedMapsMutex.Unlock()
+			// Update eventsByMessageID map
+			for _, job := range hydratedJobs {
+				msgID := job.Metadata.MessageID
+				originalJob := message.eventsByMessageID[msgID]
+				message.eventsByMessageID[msgID] = types.SingularEventWithReceivedAt{
+					SingularEvent: job.Message,
+					ReceivedAt:    originalJob.ReceivedAt,
+				}
+			}
+
+			// Update the groupedEventsBySourceId map
+			message.groupedEventsBySourceId[sourceId] = hydratedJobs
```

Comment:
If hydration failed we will be having an empty group. Do we have a test verifying the behaviour of the pipeline when a group is empty?